lib/silo/commands/software_pull.rb

```diff
@@ -60,6 +60,11 @@ def run
         end
 
         software_dir = @options.dir || Config.user_software_dir
+        cur = software_dir
+        while (!File.writable?(cur))
+          raise "User does not have permission to create files in the directory '#{cur}'" if File.exists?(cur)
+          cur = File.expand_path("..", cur)
+        end
```

Comment:
Is there a reason to have this here _and_ in `Silo#pull`?

Reply:
:duck: Because softwares are pulled to `/tmp` before they're moved to their destination.

Reply:
Beat me to it, exactly. The existing `Silo#pull` check will still check that `/tmp` is writable, but the original bug this branch was intended to fix only happened with `software pull` and not `file pull` because of this difference in implementation.

If anything, the redundant check is the one in `Silo#pull` since we _could_ assume that providers will give an appropriate error message on their own for invalid destinations, but I felt like the error messages should be consistent so I used the same definition that `software pull` now uses.